### PR TITLE
Add the git branch to the prompt string and shorten it

### PR DESCRIPTION
**Pull request by MarkDBlackwell**
Often a long prompt string is bothersome.

LilyDev's terminal emulator LXTerminal displays, in the title bar, the current username and working directory.

Without this change, the prompt string shows them, also.

This change shortens the prompt string to just the hostname and `$ '.

Now, if someone SSH's into another system, they'll still have the protection of being reminded of that fact (on every line).

This change merely comments-out the commands which had created the previous setting. This makes it easier to restore the former behavior, if someone so wishes.

### CODE REVIEW

**Comment by fedelibre** [comment]
@MarkDBlackwell Thanks for these contributions! Very welcomed but not expected.. I have a lot of changes in my local copy which I've not pushed yet. I will ask you to rebase your pull requests as soon as I have time to finish what I've done and pushed it. Sorry for the inconvenience!

Currently my high priority task is setting up the new bug tracker for LilyPond, as we must move from Google Code before August. I put LilyDev in standby because I had some bugs to fix (e.g. I'm using preseed.cfg to make the installation easier, but I think I hit a bug of the debian-installer). I've postponed my work on LilyDev until the the bug tracker stuff is complete.

One of the changes I introduced is an interactive script launched at firs login, which guides the user to set up Git (i.e.: ~/.gitignore) and download all the git repositories; then I decided to include in this script also all the stuff to be appended to .bashrc and I removed .bashrc from /etc/skel. I thought it was a good choice for maintenance, as .bashrc may change when upgrading to a new release of Debian.

**Comment by MarkDBlackwell** [comment]
> Currently my high priority task is setting up the new bug tracker for LilyPond.
> I'll ask you to rebase your pull requests as soon as I have time to finish what I've done. I have a lot of changes in my local copy.

I can understand that—no problem. :)

BTW, I see you've been involved with LilyDev since at least as far back as February 23, 2014 (on  code.google.com) with this interesting [post](https://code.google.com/p/lilypond/issues/detail?id=2538#c26), and at least as far back as September 20, 2014 (on Nabble) with this helpful [post](http://lilypond.1069038.n5.nabble.com/LilyDev-not-accepting-paste-from-host-clipboard-tt166606.html#a166651).

I'd like to take this opportunity to thank you most warmly for all your efforts! I believe they've improved LilyDev greatly. :)

File: config/includes.chroot/etc/skel/.bashrc [hunk 1]
```diff
@@ -57,9 +57,11 @@ if [ -n "$force_color_prompt" ]; then
 fi
 
 if [ "$color_prompt" = yes ]; then
-    PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
+#   PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
+    PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\h\[\033[00m\]\[\033[01;34m\]\[\033[00m\]\$ '
```
Comment (MarkDBlackwell) [inline config/includes.chroot/etc/skel/.bashrc:61]: Paul Morris [here](http://lilypond.1069038.n5.nabble.com/Improving-the-Contributors-Guide-and-LilyDev-tp176007.html) says including the current git branch in the prompt is important.

**Comment by fedelibre** [comment]
Thanks, I'm glad to hear that!

I'm having a second thought on /etc/skel/.bashrc:
- I'd rather keep it and let the interactive script at login (not pushed yet) set up only git
- a nice way to avoid replicating the basic stuff in our .bashrc would be to source the default and then add our stuff afterwards:
  
  ```
  source /etc/skel/.bashrc
  # our stuff below
  ```

What do you think about it?

I agree on removing the user from the prompt as it's not really necessary ($ and # are enough to know if you are the regular user or you have root privileges).
I strongly disagree on removing the path to working directory. I guess that most user would find it confusing and would end up typing `pwd` several times.

The changes I made included the git branch on prompt, [as recommended on the LilyPond Contributor Guide](http://lilypond.org/doc/v2.19/Documentation/contributor/setting-up#configuring-git). I would use this:

```
source /etc/bash_completion
export PS1="\h:\w\$(__git_ps1)$ "
```

I know, this does not short much the prompt. But I would stick with what the lilypond manual recommends.
A \n after \w may help, but I don't like two lines.

**Comment by MarkDBlackwell** [comment]
BASHRC

> I'm having second thoughts on /etc/skel/.bashrc: I'd rather keep it, and let the interactive script at login (not pushed yet) set up only git.

This seems good, to me.

> A nice way to avoid replicating the basic stuff in our .bashrc would be to source the default and then add our stuff afterwards:
> 
> ``` sh
> source /etc/skel/.bashrc
> # our stuff below
> ```
> 
> What do you think about this?

Some people know where to find ~/.bashrc, and how to customize it. Therefore better IMO (and with similar effect) would be the following—is this actually what you mean?:

We should keep Debian's standard contents in `/etc/skel/.bashrc`, but append it with `source /etc/skel/our-stuff`.

Most conveniently and robustly, we can accomplish this by adding something like the following as hook script `config/hooks/append-bashrc.hook.binary`:

``` sh
#!/bin/sh
set -e
if [ -w /etc/skel/.bashrc ]
then
  cat <<'EndOfAppend' >> /etc/skel/.bashrc
source /etc/skel/our-stuff
EndOfAppend
fi
```

COMPLETION-GIT MYSTERY

> The changes I made included the git branch in the prompt, as the LilyPond Contributor Guide recommends. A `\n` after `\w` may help, but I don't like two lines.

I agree with you: two lines are bad.

> I would use this:
> 
> ``` sh
> source /etc/bash_completion
> export PS1="\h:\w\$(__git_ps1)$ "
> ```

Regarding `__git_ps1`, Debian's introductory [article](https://www.debian-administration.org/article/316/An_introduction_to_bash_completion_part_1) on the topic speaks of `/etc/bash_completion`. [Comment 8](https://www.debian-administration.org/articles/316#comment_8) (from 2008) reveals that Debian split Bash completions out of package `bash` later, to their own package.

Currently, LilyDev and Debian wheezy's `bash` both lack `/etc/bash_completion`, so the above `source` command won't work:

``` sh
$ sudo ls -ld /etc/bash_completion
ls: cannot access /etc/bash_completion: No such file or directory
```

See also the Debian page by the [Bash completion](https://bash-completion.alioth.debian.org/) developers.

One provider of `__git_ps1` would seem to be the large package, `bash-completion`; LilyDev lacks that package:

``` sh
$ dpkg-query -l bash-completion
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                              Version               Architecture          Description
+++-=================================-=====================-=====================-=======================================================================
un  bash-completion                   <none>                                      (no description available)
```

However, the fact that LilyDev lacks the `bash-completion` package IMO is a good thing, since `bash-completion`:
- Is aimed at power users (and LilyDev's target isn't power users);
- Implements a large number of changes in Bash's normal behavior; and
- Isn't installed in Debian by default—i.e., Debian (for some good reason) prefers not to impose it on its newbie users.

File `/usr/share/bash-completion/completions/git` is of interest. However, that file is located in the wrong place for Bash. (Yet, see COMPLETION-GIT SOLUTION, below.)

More precisely, LilyDev's `/etc/skel/.bashrc` looks for Bash completions in the same location as Debian's original [bash](http://ftp.us.debian.org/debian/pool/main/b/bash/bash_4.2+dfsg-0.1+deb7u3_i386.deb) package file `/etc/skel/.bashrc` does: i.e., in `/usr/share/bash-completion/bash_completion`.

However, (strangely), LilyDev lacks that directory:

``` sh
$ sudo ls -ld /usr/share/bash-completion/bash_completion
ls: cannot access /usr/share/bash-completion/bash_completion: No such file or directory
```

Instead, LilyDev has Bash completion files only in `/usr/share/bash-completion/completions`. That directory's files come from two packages:

``` sh
$ dpkg-query -S '/usr/share/bash-completion/completions/*' | sort
git: /usr/share/bash-completion/completions/git
git: /usr/share/bash-completion/completions/gitk
libglib2.0-bin: /usr/share/bash-completion/completions/gdbus
libglib2.0-bin: /usr/share/bash-completion/completions/gresource
libglib2.0-bin: /usr/share/bash-completion/completions/gsettings
```

Here are LilyDev's versions of those packages:

``` sh
$ dpkg-query -l bash dash git libglib2.0-bin > packages.out
lilydev$ cat packages.out
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                 Version                        Architecture Description
+++-====================================-==============================-============-==============================================================================
ii  bash                                 4.2+dfsg-0.1+deb7u3            i386         GNU Bourne Again SHell
ii  dash                                 0.5.7-3                        i386         POSIX-compliant shell
ii  git                                  1:1.9.1-1~bpo70+2              i386         fast, scalable, distributed revision control system
ii  libglib2.0-bin                       2.33.12+really2.32.4-5         i386         Programs for the GLib library
```

Debian's package pages reveal LilyDev's [bash](https://packages.debian.org/search?suite=all&arch=i386&exact=1&searchon=names&keywords=bash), [dash](https://packages.debian.org/search?suite=all&arch=i386&exact=1&searchon=names&keywords=dash) and [libglib2.0-bin](https://packages.debian.org/search?suite=all&arch=i386&exact=1&searchon=names&keywords=libglib2.0-bin) packages are from wheezy, and LilyDev's [git](https://packages.debian.org/search?suite=all&arch=i386&exact=1&searchon=names&keywords=git) package is from wheezy-backports.

So why do both `git` and `libglib2.0-bin` refer to `/usr/share/bash-completion/completions`, but `bash` refers instead to `/usr/share/bash-completion/bash_completion`? It's mysterious.

COMPLETION-GIT SOLUTION
File `/etc/bash_completion.d/git-prompt` is of interest. It refers to `/usr/lib/git-core/git-sh-prompt`.

Section `Programmable Completion` of [man bash](http://linux.die.net/man/1/bash) reveals Bash by default already implements certain nice kinds of functionality for command-line completion.

I suppose for each additional desired Bash completion function, LilyDev should add it, one at a time, upon request. The currently desired one is `__git_ps1`.

We can add that single function by sourcing `/etc/bash_completion.d/git-prompt`.

Then, including `$(__git_ps1)` in the PS1 prompts will work properly, as in

``` sh
PS1='${debian_chroot:+($debian_chroot)}\H:\W$(__git_ps1)\$ '
```

WORKING DIRECTORY

> I agree on removing the user from the prompt as it's not really necessary (`$` and `#` are enough to know if you are the regular user or you have root privileges).
> 
> I strongly disagree on removing the path to working directory. I guess that most user would find it confusing and would end up typing pwd several times.
> 
> ``` sh
> export PS1="\h:\w\$(__git_ps1)$ "
> ```
> 
> I know, this doesn't shorten the prompt much. But I would stick with what the LilyPond manual recommends.

BTW:
- We can show solely the basename of the working directory, if we use `\W` instead of `\w`;
- The 'prompt string' which appears in LXTerminal's title bar is (completely) separately configurable from the command line's prompt string.

Perhaps the LilyDev users (in general) would find it helpful to see various information in the LXTerminal title bar, and wouldn't feel they need to see it also in the command line prompt, where it fills the line. Does this seem desirable?

**Comment by fedelibre** [comment]
> Some people know where to find ~/.bashrc, and how to customize it. 
> Therefore better IMO (and with similar effect) would be the 
> following—is this actually what you mean?:
> 
> We should keep Debian's standard contents in /etc/skel/.bashrc, but 
> append it with source /etc/skel/our-stuff.
> 
> Most conveniently and robustly, we can accomplish this by adding 
> something like the following as hook script 
> config/hooks/append-bashrc.hook.binary:
> 
> #!/bin/sh
> set -e
> if [ -w /etc/skel/.bashrc ]
> then
>   cat <<'EndOfAppend' >> /etc/skel/.bashrc
> source /etc/skel/our-stuff
> EndOfAppend
> fi

As I commented in another pull request, the binary hooks run outside 
the chroot so they should not change absolute path files or they will 
change the system running live-build.

I like very much the idea of placing our stuff in something like 
/etc/skel/.lilypond-bashrc. But we should find another way to append it 
to the default .bashrc. I see 2 options:
1. add a line in .lilydev-setup.sh (more on that below)
2. use preseed.cfg "d-i preseed/late_command" (never used, not sure it 
   will work)

> > I would use this:
> > 
> > source /etc/bash_completion
> > export PS1="\h:\w\$(__git_ps1)$ "

I'd rather stick to it. This won't be Ok for everybody but it's very 
easy to change.

> Regarding __git_ps1, Debian's introductory article on the topic 
> speaks of /etc/bash_completion. Comment 8 (from 2008) reveals that 
> Debian split Bash completions out of package bash later, to their own 
> package.
> 
> Currently, LilyDev and Debian wheezy's bash both lack 
> /etc/bash_completion, so the above source command won't work:
> 
> $ sudo ls -ld /etc/bash_completion
> ls: cannot access /etc/bash_completion: No such file or directory

Please test any change on the current LilyDev, which runs Debian Jessie 
and it does have bash_completion.

> However, the fact that LilyDev lacks the bash-completion package IMO 
> is a good thing, since bash-completion:
> 
> Is aimed at power users (and LilyDev's target isn't power users);
> Implements a large number of changes in Bash's normal behavior; and
> Isn't installed in Debian by default—i.e., Debian (for some good 
> reason) prefers not to impose it on its newbie users.

I've used bash-completion for a long time and never had any problem.

> WORKING DIRECTORY
> 
> > I agree on removing the user from the prompt as it's not really 
> > necessary ($ and # are enough to know if you are the regular user or 
> > you have root privileges).
> > 
> > I strongly disagree on removing the path to working directory. I 
> > guess that most user would find it confusing and would end up typing 
> > pwd several times.
> > 
> > export PS1="\h:\w\$(__git_ps1)$ "
> > I know, this doesn't shorten the prompt much. But I would stick with 
> > what the LilyPond manual recommends.
> > 
> > BTW:
> 
> We can show solely the basename of the working directory, if we use 
> \W instead of \w;
> The 'prompt string' which appears in LXTerminal's title bar is 
> (completely) separately configurable from the command line's prompt 
> string.
> Perhaps the LilyDev users (in general) would find it helpful to see 
> various information in the LXTerminal title bar, and wouldn't feel 
> they need to see it also in the command line prompt, where it fills 
> the line. Does this seem desirable?
> 
> I don't know: everybody has different ideas.

Perhaps we can let each user choose and add a setting to the new 
interactive script launched at first login:
https://github.com/fedelibre/LilyDev/blob/master/config/includes.chroot/etc/skel/.lilydev-setup.sh

**Comment by MarkDBlackwell** [comment]
> I'm having a second thought on `/etc/skel/.bashrc`: I'd rather keep it.
> I like very much the idea of placing our stuff in something like `/etc/skel/.lilypond-bashrc`. But we should find another way to append it to the default `~/.bashrc`. I see 2 options:
> 1. add a line in `.lilydev-setup.sh` (more on that below); and
> 2. use `preseed.cfg "d-i preseed/late_command"` (never used, not sure it will work).

I haven't looked into option 2 yet. Anyway, I would distinguish two different, useful functionalities:

\1. Easily setting up Git after reinstalling LilyDev.

This happens once automatically after installation.

BTW, IMO the CG should mention how to invoke this explicitly also, because a user might feel sometimes that he or she has made a mistake, and might immediately wish to repeat that step. Also, they might not yet at that moment wish to configure Git, but might plan instead to do so later.

\2. Customizing what happens on every login, and on every creation of a LXTerminal window.

> > `export PS1="\h:\w\$(__git_ps1)$ "`
> 
> I'd rather stick to it.
> This won't be okay for everybody but it's very easy to change.

Here, do you mean the default contents of `~/.bashrc`?

> I don't know: everybody has different ideas. Perhaps we can let each user choose and add a setting to the new interactive script launched at first login.
> `rm ~/.config/autostart/lilydev.desktop`

Of course, `~/.bashrc` is processed at each of the times mentioned in 2. Is there another such file?

If a setting in an interactive script, launched at first login, actually changes what happens in future logins (as well as future creations of LXTerminal windows), then `~/.bashrc` (or another file read at those times) must somehow incorporate the new setting—at least, that's what I suppose.

Probably LilyDev will be updated occasionally. The CG could lessen the friction for users to update LilyDev by mentioning that an individual's customizations of `~/.bashrc` (or other hypothetical login file) can be stored in a certain file (such as `~/.bashrc_personal`) outside LilyDev. Then after each reinstallation of LilyDev such user could copy in that file from outside the VM, from a place where they're keeping it safe. That would ease updating LilyDev for users who customize their login. Yet users without knowledge of this facility need do nothing.

To implement this concept, LilyDev would append merely a line to LilyDev's skeleton `~/.bashrc` to source the `~/.bashrc_personal` file, only if it exists. A script aiming at this intent is:

``` sh
if [ -w ~/.bashrc ]
then
  cat << 'EndOfAppend' >> ~/.bashrc
# Source LilyDev user's personal ~/.bashrc settings
if [ -r ~/.bashrc_personal ]
then
  source ~/.bashrc_personal
fi
EndOfAppend
fi
```

What do you think?

**Comment by fedelibre** [comment]
I've tried preseed.cfg but it's [more compicated than expected](https://lists.debian.org/debian-live/2015/08/msg00064.html).
I've few time and need to finish the new version asap. I think that I'll use an hacky solution now and come back to something better later when I have more time.

**Comment by fedelibre** [comment]
Back to the main subject of this pull: I recently discovered Fedora and I like very much the way the prompt string AND the title bar of gnome-terminal give all the information in the shortest possible space. It's very similar to what you described some messages above..

Would you like to submit a patch for this?

**Comment by fedelibre** [comment]
I've noticed that gnome-terminal title bar does not depend on .bashrc, while lxterminal does. This means that we should add the new PS1 _before_ these default lines:

``` bash
# If this is an xterm set the title to user@host:dir
case "$TERM" in
xterm*|rxvt*)
    PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
    ;;
*)
    ;;
esac
```

So we cannot append the new PS1 to the default .bashrc.
I'm afraid that we must go back to adding .bashrc in packages-list. See commit 9be871712e8c05de4870d2d04fa8959569611b1b

What do you think about it?

This would be the line I'd like to use:

```
export PS1='[\W$(declare -F __git_ps1 &>/dev/null && __git_ps1 " (%s)")]\$ '
```

**Comment by fedelibre** [comment]
fixed in commit 17d2452d5c2ac36bc2d15063fcef1331bd796153

**Pull request closed without merging**